### ISSUE TITLE
Unpacking a |%h.Map has a different behaviour from 2019.07.1

### ISSUE THREAD

**Issue by destroycomputers** [rakudo/rakudo#3168]
## The Problem
This bit of code executes differently in 2019.07.1 and HEAD. 
```
class A { has @.xs }
my %args = xs => [1, 2, 3];
A.new(|%args.Map).say
```
Relevant discussion: https://colabti.org/irclogger/irclogger_log/perl6?date=2019-09-05#l54
Commit that introduced the change in behaviour (taken from the irc log): https://github.com/rakudo/rakudo/commit/2ee82afe6e6713c26b929e2855ea05d8eb3b9fee
## Expected Behavior
The output to be `A.new(xs => [1, 2, 3])`
## Actual Behavior
Output is `A.new(xs => [[1, 2, 3],])`
## Steps to Reproduce
Run the program above.

## Environment
* Operating system: Whatever evalable6 runs on
* Compiler version (`perl6 -v`): version used by evalable6, apparently HEAD

**Comment by AlexDaniel** [comment]
Output on all releases: https://gist.github.com/f530b87b947ae95217adb0c9db6a36ac

**Comment by vrurg** [comment]
This was intentional change fixing a bug in previous releases. The fix is about synchronizing `List` and `Map` behaviors. Here is an example:

```
my $foo = 42;
my @a is List = 1, $foo;
say @a.map: { $_.VAR.^name };
my %m is Map = a => 1, b => $foo;
say %m<a b>.map: { $_.VAR.^name }
```

2019.08 output:

```
(Int Scalar)
(Int Int)
```

`HEAD` output:

```
(Int Scalar)
(Int Scalar)
```

What happens with `%h.Map` is about the fact that hashes are implicitly containerizing their values:

```
my @a = <1 2>; 
say @a.VAR.^name; # Array
my %h = a => @a; 
say %h<a>.VAR.^name; # Scalar
```

`Map` was implicitly decontainerizing its values, as the code above displays. But it doesn't anymore. Instead it preserves everything as it gets it. In our case, it preserves `Scalar`s from the hash. What actually happens with the clas initialization is something like:

```
my @a = <1 2>;
my $a = @a;
my @b = $a;
say @b; # [[1, 2]]
```

This is nothing else but a downside of using containers. The following code:

```
class A { has @.a; }; 
my %h = a => [1,2]; 
A.new(|%h.map( { $_.key => $_.value<> } ).Map).say
```

works as expected.

Perhaps what must really be done is deconting of assignments to `Array` like it currently happens with hashes:

```
my %h = a=>1, b=>2;
my $h = %h;
my %b = $h;
say %b; # {a => 1, b => 2}
```

Because this is another weird discrepancy we currently have. Perhaps a problem-solving needed on this?

**Comment by vrurg** [comment]
Ah, and one more thing: I would be protecting the current `Map` implementation for any changes because it is currently the only way to have proper way to export via `EXPORT` routines.

**Comment by AlexDaniel** [comment]
> Because this is another weird discrepancy we currently have. Perhaps a problem-solving needed on this?

Let's assume that the current behavior is right. How can you get `A.new(xs => [1, 2, 3])` output?

Previously this used to work:
```perl6
class A { has @.xs }; my %args = xs => [1, 2, 3]; A.new(|%args.Map).say
```
How to get the same behavior now?

**Comment by jnthn** [comment]
>  The fix is about synchronizing List and Map behaviors.

Indeed, but coercion of an `Array` to `.List` does actually remove the containers.

**Comment by lizmat** [comment]
Do you mean that's ok like it is, or that it should be fixed to *not* remove the containers?

**Comment by vrurg** [comment]
Because the discussion heading a bit out of the scope of this ticket, I have created perl6/problem-solving#99.

**Comment by vrurg** [comment]
Proposed fix implementation is in rakudo/rakudo#3174, tests are in perl6/roast#575.

**Comment by vrurg** [comment]
Closing via #3174
